cupy/core/_reduction.pyx

```diff
@@ -290,13 +299,118 @@ cdef class _AbstractReductionKernel:
+            if params is None:
+                params = self._get_optimized_params(
+                    optimize_context.config, in_args, out_args,
+                    in_shape, out_shape, type_map, map_expr, reduce_expr,
+                    post_map_expr, reduce_type, stream)
+                optimize_context.set_params(key, params)
+            block_size, block_stride, out_block_num = params
+
+        # Launch the kernel
+        self._launch(
+            out_block_num,
+            block_size,
+            block_stride,
+            in_args, out_args,
+            in_shape, out_shape,
+            type_map,
+            map_expr, reduce_expr, post_map_expr, reduce_type,
+            stream)
+        return ret
+
+    def _get_optimized_params(
+            self, optimize_config, in_args, out_args, in_shape, out_shape,
+            type_map, map_expr, reduce_expr, post_map_expr, reduce_type,
+            stream):
+        out_size = internal.prod_sequence(out_shape)
+
+        def copy_arg(a):
+            if isinstance(a, ndarray):
+                x = _create_ndarray_from_shape_strides(
+                    a._shape, a._strides, a.dtype)
+                assert a.data.device_id == x.data.device_id
+                x[:] = a
+                return x
+            return a
+
+        in_args = [copy_arg(a) for a in in_args]
+        out_args = [copy_arg(a) for a in out_args]
+
         contiguous_size = _get_contiguous_size(
-            in_args, self.in_params, in_shape.size(), out_shape.size())
-        block_size, block_stride, out_block_num = _get_block_specs(
+            in_args, self.in_params, len(in_shape), len(out_shape))
+        block_size, block_stride, default_out_block_num = _get_block_specs(
             internal.prod(in_shape),
             internal.prod(out_shape),
-            contiguous_size)
+            contiguous_size, -1)
+        default_block_size_log = math.floor(math.log2(block_size))
+        default_block_stride_log = math.floor(math.log2(block_size))
+
+        def target_func(block_size, block_stride, out_block_num):
+            self._launch(
+                out_block_num, block_size, block_stride, in_args, out_args,
+                in_shape, out_shape, type_map, map_expr, reduce_expr,
+                post_map_expr, reduce_type, stream)
+
+        def suggest_func(trial):
+            block_size_log = trial.suggest_int('block_size_log', 5, 9)
+            block_size = 2 ** block_size_log
+            block_stride_log = trial.suggest_int(
+                'block_stride_log', 0, block_size_log)
+            block_stride = 2 ** block_stride_log
+            max_out_block_num = (out_size + block_stride - 1) // block_stride
+            out_block_num = trial.suggest_int(
+                'out_block_num', 1, max_out_block_num)
+
+            trial.set_user_attr('block_size', block_size)
+            trial.set_user_attr('block_stride', block_stride)
+            return block_size, block_stride, out_block_num
```

Comment:
Hi @asi1024 May I ask why `out_block_num` isn't set as a user attribute here?

Reply:
The value of `out_block_num` is already registered [here](https://github.com/cupy/cupy/pull/2731/files#diff-f28653b5a69ca2e5bf585b8dd3e9dd02R388-R389) whereas `block_size` and `block_stride` are not. (`block_size_log` and `block_stride_log` are registered.)
note: [Optuna User Attributes](https://optuna.readthedocs.io/en/latest/tutorial/attributes.html#user-attributes)

Reply:
Ah, I see. So you meant only those properties not registered should be set as user attributes?

Reply:
We can write [here](https://github.com/cupy/cupy/pull/2731/files#diff-f28653b5a69ca2e5bf585b8dd3e9dd02R405-R407) as
```
2 ** best.params['block_size_log'],
2 ** best.params['block_stride_log'],
best.params['out_block_num']
```
without setting user attributes, but I feel the current implementation is better.

Reply:
I see, good to know! I pinged you in my PR for further discussion.